sql/tests: TestRandomSyntaxFunctions failed [panic in request_job_execution_details]

sql/tests.TestRandomSyntaxFunctions [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/10950443?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/10950443?buildTab=artifacts#/) on master @ [7675ca4998134028f0623e04737b5cb69fcc33a9](https://github.com/cockroachdb/cockroach/commits/7675ca4998134028f0623e04737b5cb69fcc33a9):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 4593578 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:479
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval
*   | 	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sem/tree/eval_expr_generated.go:281
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
*   | github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec
*   | 	github.com/cockroachdb/cockroach/pkg/sql/values.go:84
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan.go:519
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan.go:522
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:725
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:575
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:301
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1955
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1669
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1672
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2097
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1656
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:964
* Wraps: (4) runtime error: index out of range [0] with length 0
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.boundsError
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #99182 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-23.1]
- #95619 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
- #95617 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
- #95616 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFunctions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29780

sql/tests.TestRandomSyntaxFunctions [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/10967177?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/10967177?buildTab=artifacts#/) on master @ [06b5ba7888339ed53477ff73f9f3ae77a708ee17](https://github.com/cockroachdb/cockroach/commits/06b5ba7888339ed53477ff73f9f3ae77a708ee17):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 3706228 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:479
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval
*   | 	github.com/cockroachdb/cockroach/bazel-out/k8-dbg/bin/pkg/sql/sem/tree/eval_expr_generated.go:281
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
*   | github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec
*   | 	github.com/cockroachdb/cockroach/pkg/sql/values.go:84
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan.go:519
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan.go:522
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:725
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:575
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:301
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1955
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1669
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1672
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2097
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1656
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:964
* Wraps: (4) runtime error: index out of range [0] with length 0
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.boundsError
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #99182 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-23.1]
- #95619 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
- #95617 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
- #95616 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFunctions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/jobs/utils.go
+++ b/pkg/jobs/utils.go
@@ -17,6 +17,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/errors"
@@ -139,4 +140,16 @@ ORDER BY created`
 		}
 	}
 	return false /* exists */, err
+}
+
+// JobExists returns true if there is a row corresponding to jobID in the
+// system.jobs table.
+func JobExists(
+	ctx context.Context, jobID jobspb.JobID, txn *kv.Txn, ex isql.Executor,
+) (bool, error) {
+	row, err := ex.QueryRow(ctx, "check-for-job", txn, `SELECT id FROM system.jobs WHERE id = $1`, jobID)
+	if err != nil {
+		return false, err
+	}
+	return row != nil, nil
 }

--- a/pkg/sql/jobs_profiler_execution_details.go
+++ b/pkg/sql/jobs_profiler_execution_details.go
@@ -176,12 +176,22 @@ func constructBackupExecutionDetails(
 // RequestExecutionDetailFiles implements the JobProfiler interface.
 func (p *planner) RequestExecutionDetailFiles(ctx context.Context, jobID jobspb.JobID) error {
 	execCfg := p.ExecCfg()
-	if !execCfg.Settings.Version.IsActive(ctx, clusterversion.V23_1) {
+	if !execCfg.Settings.Version.IsActive(ctx, clusterversion.V23_2) {
 		return errors.Newf("execution details can only be requested on a cluster with version >= %s",
-			clusterversion.V23_1.String())
+			clusterversion.V23_2.String())
 	}
 
 	e := makeJobProfilerExecutionDetailsBuilder(execCfg.SQLStatusServer, execCfg.InternalDB, jobID)
+
+	// Check if the job exists otherwise we can bail early.
+	exists, err := jobs.JobExists(ctx, jobID, p.Txn(), e.db.Executor())
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return errors.Newf("job %d not found; cannot request execution details", jobID)
+	}
+
 	// TODO(adityamaru): When we start collecting more information we can consider
 	// parallelize the collection of the various pieces.
 	e.addDistSQLDiagram(ctx)
@@ -238,7 +248,7 @@ func (e *executionDetailsBuilder) addDistSQLDiagram(ctx context.Context) {
 		log.Errorf(ctx, "failed to write DistSQL diagram for job %d: %+v", e.jobID, err.Error())
 		return
 	}
-	if row[0] != tree.DNull {
+	if row != nil && row[0] != tree.DNull {
 		dspDiagramURL := string(tree.MustBeDString(row[0]))
 		filename := fmt.Sprintf("distsql.%s.html", timeutil.Now().Format("20060102_150405.00"))
 		if err := jobs.WriteExecutionDetailFile(ctx, filename,

--- a/pkg/sql/jobs_profiler_execution_details_test.go
+++ b/pkg/sql/jobs_profiler_execution_details_test.go
@@ -220,6 +220,10 @@ func TestReadWriteProfilerExecutionDetails(t *testing.T) {
 		require.True(t, strings.Contains(string(goroutines), fmt.Sprintf("labels: {\"foo\":\"bar\", \"job\":\"IMPORT id=%d\", \"n\":\"1\"}", importJobID)))
 		require.True(t, strings.Contains(string(goroutines), "github.com/cockroachdb/cockroach/pkg/sql_test.fakeExecResumer.Resume"))
 	})
+
+	t.Run("execution details for invalid job ID", func(t *testing.T) {
+		runner.ExpectErr(t, `job -123 not found; cannot request execution details`, `SELECT crdb_internal.request_job_execution_details(-123)`)
+	})
 }
 
 func TestListProfilerExecutionDetails(t *testing.T) {

